File: src/ustplayer/core/settings_manager.py
# settings_manager.py — 应用设置门面（组装各设置子域）
"""SettingsManager 只负责组装与编排，不再持有具体配置属性：

- 设置子域（属性 + 信号 + 分组读写）分布在 core/settings/ 包：
  ProjectSettings / FileSettings / DisplaySettings / ColorSettings / PlayerSettings / ThemeSettings；
- Settings.json 文件存取（含旧版 Settings.ini 自动迁移）由 core/settings_store.py（SettingsStore）承担；
- .uplr 导入/导出由 core/uplr_io.py（UplrProjectIO）承担。

UI 通过 ctx.settings.<子域>.<属性> 访问，如 ctx.settings.display.show_bpm。
"""

import os
from typing import Dict, Optional

from PySide6.QtCore import QObject

from ustplayer.core.contracts import (
    PlayerLaunchParams,
    PlayerStyle,
    ProjectInfo,
    ShowConfig,
    UstInfo,
    resolve_program_root,
)
from ustplayer.core.log import logger
from ustplayer.core.settings import (
    ColorSettings,
    DisplaySettings,
    FileSettings,
    LanguageSettings,
    PlayerSettings,
    ProjectSettings,
    ThemeSettings,
)
from ustplayer.core.settings_store import SettingsStore


class SettingsManager(QObject):
    """应用设置门面 — 组装六个设置子域并编排 ini 读写/校验/参数组装。"""

    def __init__(self, parent: Optional[QObject] = None):
        super().__init__(parent)

        self.program_root = resolve_program_root()
        self._store = SettingsStore()
        self.settings_path = self._store.settings_path

        self.terms_file_path = os.path.join(self.program_root, "LICENSE")
        self.ercode_file_path = os.path.join(self.program_root, "ERcode.txt")

        default_desktop = os.path.join(os.path.expanduser("~"), "Desktop")
        self.last_open_dir = default_desktop
        self.last_export_dir = default_desktop

        self.project = ProjectSettings(self)
        self.file = FileSettings(self)
        self.display = DisplaySettings(self)
        self.color = ColorSettings(self)
        self.player = PlayerSettings(self)
        self.theme = ThemeSettings(self)
        self.language = LanguageSettings(self)

        self._config: Dict[str, Dict[str, str]] = {}
        self.read_settings()

    # ===================== Settings.json 读写 =====================

    def read_settings(self):
        """读取设置并恢复全部配置，路径失效时回退桌面。"""
        default_desktop = os.path.join(os.path.expanduser("~"), "Desktop")
        try:
            self._config = self._store.load()
            if not self._config:
                self.last_open_dir = default_desktop
                self.last_export_dir = default_desktop
                return

            if "PathSettings" in self._config:
                self.last_open_dir = self._config["PathSettings"].get(
                    "last_open_dir", default_desktop
                )
                self.last_export_dir = self._config["PathSettings"].get(
                    "last_export_dir", default_desktop
                )
                if not os.path.isdir(self.last_open_dir):
                    self.last_open_dir = default_desktop
                if not os.path.isdir(self.last_export_dir):
                    self.last_export_dir = default_desktop

            self.project.read_from(self._config)
            self.file.read_from(self._config)
            self.display.read_from(self._config)
            self.color.read_from(self._config)
            self.player.read_from(self._config)
            self.theme.read_from(self._config)
            self.language.read_from(self._config)

            self.sanitize()
            self.write_settings()
        except Exception as e:
            self.last_open_dir = default_desktop
            self.last_export_dir = default_desktop
            logger.exception(f"读取配置文件失败：{e}")

    def write_settings(self):
        """将所有设置写入配置文件，退出时保存以便重启恢复。"""
        try:
            self._config["PathSettings"] = {
                "last_open_dir": self.last_open_dir,
                "last_export_dir": self.last_export_dir,
            }
            self.project.write_to(self._config)
            self.file.write_to(self._config)
            self.display.write_to(self._config)
            self.color.write_to(self._config)
            self.player.write_to(self._config)
            self.theme.write_to(self._config)
            self.language.write_to(self._config)

            self._store.save(self._config)
        except Exception as e:
            logger.exception(f"写入配置文件失败：{e}")

    # ===================== 校验 =====================

    def sanitize(self):
        """校验各子域的枚举/颜色值，非法时回退默认。"""
        self.file.validate()
        self.color.validate()
        self.player.validate()

    # ===================== 构建播放器启动参数 =====================

    def build_ust_info(self, core_ust_info: UstInfo) -> PlayerLaunchParams:
        """将解析结果与当前设置组装为播放器启动参数（统一接口）。"""
        return PlayerLaunchParams(
            ust=core_ust_info,
            show=ShowConfig(
                bpm=self.display.show_bpm,
                play_time=self.display.show_play_time,
                song_name=self.display.show_song_name,
                song_author=self.display.show_song_author,
                ust_author=self.display.show_ust_author,
                lyric=self.display.show_lyric,
                curve_show=self.file.curve_show,
            ),
            project=ProjectInfo(
                project_name=self.project.project_name,
                song_name=self.project.song_name,
                song_author=self.project.song_author,
                ust_author=self.project.ust_author,
            ),
            style=PlayerStyle(
                bg_color=self.color.bg_color,
                note_color=self.color.note_color,
                lyric_color=self.color.lyric_color,
                lyric_text_color=self.color.lyric_text_color,
                other_text_color=self.color.other_text_color,
                lyric_pos=self.player.lyric_pos,
                fullscreen=self.display.fullscreen,
                lrc_path=self.player.lrc_path,
                music_path=self.project.music_path,
                silent_display=self.player.silent_display,
                silent_custom_text=self.player.silent_custom_text,
                end_display=self.player.end_display,
                end_custom_text=self.player.end_custom_text,
                pitch_placeholder=self.player.pitch_placeholder,
                pitch_custom_text=self.player.pitch_custom_text,
                pitch_curve_color=self.color.pitch_curve_color,
            ),
        )

File: ustPlayer/Settings/SettingsManager.cs
using System;
using System.IO;

using UstPlayer.Diagnostics;
using UstPlayer.Models;
using UstPlayer.Settings;
using UstPlayer.Settings.Domains;

namespace UstPlayer.Settings;

/// <summary>
/// 设置门面 — 从 1.1.x <c>core/settings_manager.py</c> 移植。
/// </summary>
/// <remarks>
/// <para>
/// 只负责**组装**七个设置子域并编排读写，自身不持有任何配置属性。
/// UI 通过 <c>ctx.settings.Display.ShowBpm</c> 这类路径访问。
/// </para>
/// <para>
/// 与 1.1.x 的差异：设置文件路径由本类持有（1.1.x 由 <c>SettingsStore</c> 内部决定），
/// 便于测试注入与将来的便携模式；<c>program_root</c> 改为读 <see cref="ProgramPaths"/>。
/// </para>
/// </remarks>
internal sealed class SettingsManager
{
    private readonly SettingsStore _store;

    /// <summary>创建设置管理器并立即读取设置。</summary>
    /// <param name="settingsPath">设置文件路径；传 <see langword="null"/> 用默认策略解析。</param>
    internal SettingsManager(string? settingsPath = null)
    {
        _store = new SettingsStore(settingsPath);

        Project = new ProjectSettings();
        File = new FileSettings();
        Display = new DisplaySettings();
        Color = new ColorSettings();
        Player = new PlayerSettings();
        Language = new LanguageSettings();
        Theme = new ThemeSettings();

        // 上次打开的目录默认指向桌面（与 1.1.x 一致）
        var desktop = Environment.GetFolderPath(Environment.SpecialFolder.DesktopDirectory);
        LastOpenDirectory = desktop;
        LastExportDirectory = desktop;

        ReadSettings();
    }

    /// <summary>项目信息子域。</summary>
    internal ProjectSettings Project { get; }

    /// <summary>文件与编码子域。</summary>
    internal FileSettings File { get; }

    /// <summary>显示开关与字体子域。</summary>
    internal DisplaySettings Display { get; }

    /// <summary>颜色子域。</summary>
    internal ColorSettings Color { get; }

    /// <summary>播放器样式子域。</summary>
    internal PlayerSettings Player { get; }

    /// <summary>语言子域。</summary>
    internal LanguageSettings Language { get; }

    /// <summary>主题子域。</summary>
    internal ThemeSettings Theme { get; }

    /// <summary>程序根目录（缓存目录与内置资源都相对它解析）。</summary>
    internal string ProgramRoot => ProgramPaths.ProgramRoot;

    /// <summary>当前设置文件路径。</summary>
    internal string SettingsPath => _store.SettingsPath;

    /// <summary>上次打开工程的目录。</summary>
    internal string LastOpenDirectory { get; set; }

    /// <summary>上次导出视频的目录。</summary>
    internal string LastExportDirectory { get; set; }

    /// <summary>
    /// 读取设置并恢复全部配置。
    /// </summary>
    /// <remarks>
    /// 读取失败（文件损坏等）不抛异常：回退默认值并记录日志——坏掉的配置文件
    /// 不该让程序无法启动（与 1.1.x 一致）。
    /// </remarks>
    internal void ReadSettings()
    {
        var desktop = Environment.GetFolderPath(Environment.SpecialFolder.DesktopDirectory);

        try
        {
            var config = _store.Load();

            if (config.IsEmpty)
            {
                LastOpenDirectory = desktop;
                LastExportDirectory = desktop;
                return;
            }

            var pathGroup = config.GetSection(SettingsSections.Path);
            if (pathGroup is not null)
            {
                var open = pathGroup.GetNode("last_open_dir") is { } openNode &&
                           openNode.GetValueKind() == System.Text.Json.JsonValueKind.String
                    ? openNode.GetValue<string>()
                    : desktop;

                var export = pathGroup.GetNode("last_export_dir") is { } exportNode &&
                             exportNode.GetValueKind() == System.Text.Json.JsonValueKind.String
                    ? exportNode.GetValue<string>()
                    : desktop;

                // 目录已不存在（U 盘拔掉、目录改名）时回退桌面
                LastOpenDirectory = Directory.Exists(open) ? open : desktop;
                LastExportDirectory = Directory.Exists(export) ? export : desktop;
            }

            Project.ReadFrom(config);
            File.ReadFrom(config);
            Display.ReadFrom(config);
            Color.ReadFrom(config);
            Player.ReadFrom(config);
            Language.ReadFrom(config);
            Theme.ReadFrom(config);
        }
        catch (Exception exception)
        {
            LastOpenDirectory = desktop;
            LastExportDirectory = desktop;
            AppLogger.Error("读取配置文件失败，使用默认值", exception);
        }
    }

    /// <summary>把全部设置写回配置文件（退出时保存，以便重启恢复）。</summary>
    internal void WriteSettings()
    {
        try
        {
            var config = new SettingsConfig(new System.Text.Json.Nodes.JsonObject());

            var pathGroup = config.GetSection(SettingsSections.Path, createIfMissing: true)!;
            pathGroup.SetString("last_open_dir", LastOpenDirectory);
            pathGroup.SetString("last_export_dir", LastExportDirectory);

            Project.WriteTo(config);
            File.WriteTo(config);
            Display.WriteTo(config);
            Color.WriteTo(config);
            Player.WriteTo(config);
            Language.WriteTo(config);
            Theme.WriteTo(config);

            _store.Save(config);
        }
        catch (Exception exception)
        {
            AppLogger.Error("写入配置文件失败", exception);
        }
    }

    /// <summary>
    /// 组装播放器的启动参数。
    /// </summary>
    /// <param name="ustInfo">UST 解析结果。</param>
    /// <returns>播放器启动参数。</returns>
    /// <remarks>
    /// 注意 <c>curve_show</c> 取自 <see cref="FileSettings"/>（1.1.x 即如此），
    /// 而非显示子域。
    /// </remarks>
    internal PlayerLaunchParams BuildLaunchParams(UstInfo ustInfo)
    {
        ArgumentNullException.ThrowIfNull(ustInfo);

        return new PlayerLaunchParams
        {
            Ust = ustInfo,
            Show = new ShowConfig
            {
                Bpm = Display.ShowBpm,
                PlayTime = Display.ShowPlayTime,
                SongName = Display.ShowSongName,
                SongAuthor = Display.ShowSongAuthor,
                UstAuthor = Display.ShowUstAuthor,
                Lyric = Display.ShowLyric,
                CurveShow = File.CurveShow,
                NoteName = Display.ShowNoteName,
                UstLyric = Display.ShowUstLyric,
                Copyright = Display.ShowCopyright,
                FontNote = Display.FontNote,
                FontUstLyric = Display.FontUstLyric,
                FontLrc = Display.FontLrc,
                FontOther = Display.FontOther,
                CustomFontPaths = [.. Display.CustomFontPaths],
            },
            Project = new ProjectInfo
            {
                ProjectName = Project.ProjectName,
                SongName = Project.SongName,
                SongAuthor = Project.SongAuthor,
                UstAuthor = Project.UstAuthor,
            },
            Style = new PlayerStyle
            {
                BackgroundColor = Color.BackgroundColor,
                NoteColor = Color.NoteColor,
                LyricColor = Color.LyricColor,
                LyricTextColor = Color.LyricTextColor,
                OtherTextColor = Color.OtherTextColor,
                LyricPosition = Player.LyricPosition,
                Fullscreen = Display.Fullscreen,
                LrcPath = Player.LrcPath,
                MusicPath = Project.MusicPath,
                SilentDisplay = Player.SilentDisplay,
                SilentCustomText = Player.SilentCustomText,
                EndDisplay = Player.EndDisplay,
                EndCustomText = Player.EndCustomText,
                PitchPlaceholder = Player.PitchPlaceholder,
                PitchCustomText = Player.PitchCustomText,
                PitchCurveColor = Color.PitchCurveColor,
                AppVersion = AppInfo.Version,
            },
        };
    }

    /// <summary>
    /// 快照「会被工程导入触碰」的全部设置属性，供失败回滚。
    /// </summary>
    /// <returns>快照。</returns>
    internal SettingsSnapshot CreateSnapshot() => SettingsSnapshot.Capture(this);

    /// <summary>按快照恢复设置（setter 会触发通知，UI 因此同步回旧值）。</summary>
    /// <param name="snapshot">快照。</param>
    internal void RestoreSnapshot(SettingsSnapshot snapshot)
    {
        ArgumentNullException.ThrowIfNull(snapshot);
        snapshot.Restore();
    }
}
